### PR TITLE
ARROW-13441: [C++][CSV] Skip empty batches in column decoder

### PR DESCRIPTION
**Pull request by n3world**
When the infering column decoder encounters an empty batch just return an empty array immediatly and do not consider the batch as the first batch.

### CODE REVIEW

**Review by pitrou — CHANGES_REQUESTED** [review]
The problem with this approach is that yields column chunks of differing types. You can see this if you add the following test:
```patch
diff --git a/cpp/src/arrow/csv/column_decoder_test.cc b/cpp/src/arrow/csv/column_decoder_test.cc
index c8b96e046..d90509df0 100644
--- a/cpp/src/arrow/csv/column_decoder_test.cc
+++ b/cpp/src/arrow/csv/column_decoder_test.cc
@@ -303,6 +303,16 @@ class InferringColumnDecoderTest : public ColumnDecoderTest {
     AssertFetch(ArrayFromJSON(type, "[901, null]"));
   }
 
+  void TestEmptyFirstBlock() {
+    auto type = int64();
+
+    MakeDecoder(default_options);
+    AppendChunks({{}, {"123", "456", "-78"}, {"901", "N/A"}});
+    AssertFetch(ArrayFromJSON(type, "[]"));
+    AssertFetch(ArrayFromJSON(type, "[123, 456, -78]"));
+    AssertFetch(ArrayFromJSON(type, "[901, null]"));
+  }
+
   void TestThreaded() {
     constexpr int NITERS = 10;
     auto type = float64();
@@ -371,6 +381,8 @@ class InferringColumnDecoderTest : public ColumnDecoderTest {
 
 TEST_F(InferringColumnDecoderTest, Integers) { this->TestIntegers(); }
 
+TEST_F(InferringColumnDecoderTest, EmptyFirstBlock) { this->TestEmptyFirstBlock(); }
+
 TEST_F(InferringColumnDecoderTest, Threaded) { this->TestThreaded(); }
 
 TEST_F(InferringColumnDecoderTest, Options) { this->TestOptions(); }
```

You will get:
```
[ RUN      ] InferringColumnDecoderTest.EmptyFirstBlock
/home/antoine/arrow/dev/cpp/src/arrow/testing/gtest_util.cc:132: Failure
Failed
# Array types differed: int64 vs null

[  FAILED  ] InferringColumnDecoderTest.EmptyFirstBlock (0 ms)
```

**Comment by n3world** [comment]
> The problem with this approach is that yields column chunks of differing types. You can see this if you add the following test:

Yes that will because the type is unknown, yet. This test seems artificial in that it doesn't follow how the column decoder is actually used. In use all empty record batches get discarded so their type don't actually matter. That is why it works for the csv streaming test I modified to have multiple empty blocks before a block with data.

That test would not work without this change so the change does not make anything better or worse as far as that test goes. I would argue it makes things better because eventually you can get data from the converter.

The only ways I can think to get that test to work would be to allow the API not to have to return a result until the type is known but that is dependent on the read ahead, for csv, to be high enough to actually find data.

**Comment by pitrou** [comment]
> Yes that will because the type is unknown, yet. This test seems artificial in that it doesn't follow how the column decoder is actually used. In use all empty record batches get discarded so their type don't actually matter.

Ah, I see. Can you point me where this happens exactly, though?

**Comment by n3world** [comment]
> > Yes that will because the type is unknown, yet. This test seems artificial in that it doesn't follow how the column decoder is actually used. In use all empty record batches get discarded so their type don't actually matter.
> 
> Ah, I see. Can you point me where this happens exactly, though?

For the first block it happens at `StreamingReaderImpl::InitAfterFirstBatch:920-925`. Looking again at this the schema is already captured so that is an issue.

I don't seem to be able to find where it happens other than the first block. I know it use to be in the previous streaming reader, SerialStreamingReader::ReadNextSkippingEmpty. @westonpace was this an intentional change or an accident? Either way this does need a little more work in the csv reader to be able to handle consuming the empty leading blocks and capturing the schema after the first one with data.

**Comment by pitrou** [comment]
> For the first block it happens at StreamingReaderImpl::InitAfterFirstBatch:920-925. Looking again at this the schema is already captured so that is an issue.

That was my impression as well. 

In any case, it would be nice to add an explicit StreamingReader test on the C++ side for this.

**Comment by westonpace** [comment]
Yes, I removed `SerialStreamingReader::ReadNextSkippingEmpty` semi-intentionally as I thought there could only be one skipped batch.  However, I hadn't realized that the right combinations of skip settings could yield multiple empty parsed blocks so that logic doesn't work.  Filtering in parallel is tricky and rather than solve that I think we can get away with filtering on one of the serial spots.

Here is a commit that adds a filtering utility to the async generators: https://github.com/westonpace/arrow/commit/8d80722a8b8cd304dff8322e96d76dfd75899ea7

I think the cleanest might be to filter empty parsed blocks when looking for the first batch.  This should be safe (as long as you pass in rb_gen and not filtered_rb_gen to use after you got the first batch):
```
    auto filtered_rb_gen = MakeFilteredGenerator(rb_gen, std::move(not_empty));
    return filtered_rb_gen().Then([self, rb_gen, max_readahead](const DecodedBlock& first_block) {
      return self->InitAfterFirstBatch(first_block, std::move(rb_gen), max_readahead);
    });
```
Or we could filter empty batches on the other side of the readahead generator.

**Comment by n3world** [comment]
> I think the cleanest might be to filter empty parsed blocks when looking for the first batch. This should be safe (as long as you pass in rb_gen and not filtered_rb_gen to use after you got the first batch):

I did something along these lines in my last update which essentially recursively called InitAfterFirstBatch until either the RecordBatch was null or it had rows. I could probably update that approach to use this new filter if you would prefer that.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:913]:
```suggestion
      record_batch_gen_ = MakeEmptyGenerator<std::shared_ptr<RecordBatch>>();
```

File: cpp/src/arrow/csv/column_decoder.cc
Comment (westonpace) [inline cpp/src/arrow/csv/column_decoder.cc:201]:
```suggestion
  // Note: Care should be taken to discard these batches as the data type will be null
  // and won't match future decoded arrays.
  if (parser->num_rows() == 0) {
```

Nit: Not necessary but might help explain to future readers.  Optionally you could also return `nullptr` here so it is clear it is invalid but then that will cause some confusion below between EOF and invalidly parsed batch.  I'll defer to whatever you think is best.

**Review by westonpace — APPROVED** [review]
I think a recursive approach is fine, no need to change.  This LGTM.  Thanks for fixing this.  If it isn't too much trouble could you maybe add one test (python is fine) opening a file that has a header and some skipped data but is otherwise empty (maybe one case where skip after header is more than actual data and one case where it exactly equals actual data) and then verify that the schema is a nullptr/None?  I think this could be an extension to one of the cases in `test_header_skip_rows`.

File: cpp/src/arrow/csv/reader.cc
Comment (n3world) [reply cpp/src/arrow/csv/reader.cc]: Done

**Comment by n3world** [comment]
> In any case, it would be nice to add an explicit StreamingReader test on the C++ side for this.

Are the tests I added sufficient?

**Comment by westonpace** [comment]
Might be good to mention @pitrou with the question to make sure it gets noticed.

File: cpp/src/arrow/csv/reader_test.cc
Comment (pitrou) [inline cpp/src/arrow/csv/reader_test.cc:393]: You can construct an expected schema and call `AssertSchemaEqual`.

File: cpp/src/arrow/csv/reader_test.cc
Comment (pitrou) [inline cpp/src/arrow/csv/reader_test.cc:394]: Can you also test the contents of `batch`? You can use `RecordBatchFromJSON` to construct the expected record batch.

File: python/pyarrow/tests/test_csv.py
Comment (pitrou) [inline python/pyarrow/tests/test_csv.py:597]: Not fond of this reorganization at all. Can't your test fit in the existing structure?

Comment (n3world) [reply python/pyarrow/tests/test_csv.py:597]: My goal was to unify the test structure of the table and stream tests so that tests can be written once and executed for both readers. I opted to go to the parameter approach because that was recently done for the streaming tests and assumed since that was more recent it was the more desired style. I can revert it back to the inheritance approach instead but that would mean I would want to undo the parameterization that @westonpace recently did on the streaming tests.

It doesn't matter to me which approach is used but it makes it a lot easier if it is the same style for testing the different readers.

Comment (pitrou) [reply python/pyarrow/tests/test_csv.py:597]: Hmm, sorry, that's a bit of an annoyance indeed. Personally, I would rather not have additional `use_threads` arguments in every test function, so I prefer the original subclassing approach.

Comment (n3world) [reply python/pyarrow/tests/test_csv.py:597]: I do agree that the extra `use_threads` argument is a little annoying for every test. I wish the parameter was passed to the class init rather than the test method then it would only be in one spot.

File: python/pyarrow/tests/test_csv.py
Comment (pitrou) [inline python/pyarrow/tests/test_csv.py:327]: Ah. I'm sorry to insist, but I'd prefer that we avoid this kind of obscure feature/hack. Surely there is a way to make these additional tests fit with the original subclassing scheme?

Comment (westonpace) [reply python/pyarrow/tests/test_csv.py:327]: @pitrou It's fine to insist.  I feel somewhat responsible for introducing the schism in the first place so I've submitted a PR @n3world that should go back to the old subclassing scheme: https://github.com/n3world/arrow/pull/2

Comment (n3world) [reply python/pyarrow/tests/test_csv.py]: @pitrou, I added @westonpace change and simplified it a bit so that the only method actually in the test classes is the use_threads property method.

Comment (pitrou) [reply python/pyarrow/tests/test_csv.py]: Thank you very much @n3world !

**Review by pitrou — APPROVED** [review]
Thanks for being persistent @n3world. I'll merge if CI passes.

File: python/pyarrow/tests/test_csv.py
Comment (n3world) [inline python/pyarrow/tests/test_csv.py:341]: The NotImplementedError shouldn't be needed since the class cannot be created if an abstract method is not implemented

Comment (n3world) [reply python/pyarrow/tests/test_csv.py:341]: I just realized there is an `abstractproperty` decorator which I should have used here. Guess I'll fix that in my next change someday

Comment (pitrou) [reply python/pyarrow/tests/test_csv.py:341]: Hmm, you're right.

**Pull request closed without merging**
